File: 790. Domino and Tromino Tiling/790. Domino and Tromino Tiling top-down.java
class Solution {
    private long[] memo;
    private long MOD = 1_000_000_007;

    private long dp(int i) {
        if (i <= 0) {
            return 0;
        }
        if (i == 1) {
            return 1;
        }
        if (i == 2) {
            return 2;
        }
        if (i == 3) {
            return 5;
        }
        if (memo[i] == 0) {
            memo[i] = ((2 * dp(i - 1) + dp(i - 3)) % MOD);
        }
        return memo[i];
    }

    public int numTilings(int n) {
        this.memo = new long[n + 1];
        return (int) (dp(n) % MOD);
    }
}

File: 790. Domino and Tromino Tiling/790. Domino and Tromino Tiling top-down.py
from functools import lru_cache


class Solution:
    def numTilings(self, n: int) -> int:
        modulo = 10**9+7

        @lru_cache(None)
        def dp(i: int) -> int:
            if i == 0:
                return 1
            if i == 1:
                return 1
            if i == 2:
                return 2
            if i == 3:
                return 5
            return (dp(i-1)*2 + dp(i-3)) % modulo
        return dp(n)


s = Solution()
n = 6
print(s.numTilings(n))
